api/src/main/java/org/openmrs/module/pihcore/event/PatientUpdateEventConsumer.java

```diff
@@ -0,0 +1,212 @@
+
+    private final DbEventSourceConfig config;
+    private final Database database;
+    private final Map<String, String> patientKeys = new LinkedHashMap<>();
+    private Rocks statusDb;
+    private boolean snapshotInitialized = false;
+
+    public PatientUpdateEventConsumer(DbEventSourceConfig config) {
+        this.config = config;
+        this.database = config.getContext().getDatabase();
+        patientKeys.put("person_id", "person");
+        patientKeys.put("person_a", "person");
+        patientKeys.put("person_b", "person");
+        patientKeys.put("order_id", "orders");
+        patientKeys.put("patient_program_id", "patient_program");
+        patientKeys.put("encounter_id", "encounter");
+        patientKeys.put("visit_id", "visit");
+        patientKeys.put("allergy_id", "allergy");
+        patientKeys.put("appointment_id", "appointmentscheduling_appointment");
+    }
+
+    @Override
+    public void startup() {
+        try {
+            statusDb = new Rocks(new File(config.getContext().getModuleDataDir(), "status.db"));
+            snapshotInitialized = BooleanUtils.isTrue(statusDb.get("snapshotInitialized"));
+        }
+        catch (Exception e) {
+            throw new RuntimeException("An error occurred starting up", e);
+        }
+    }
+
+    @Override
+    public void shutdown() {
+        statusDb.close();
+    }
+
+    @Override
+    public void accept(DbEvent event) {
+        if (!snapshotInitialized) {
+            performInitialSnapshot();
+            statusDb.put("snapshotInitialized", Boolean.TRUE);
+            snapshotInitialized = true;
+        }
+        Set<Integer> patientIds = getPatientIdsForEvent(event);
+        if (patientIds.isEmpty()) {
+            log.warn("Not handling event as not mapped to patient: " + event);
+        }
+        else {
+            for (Integer patientId : patientIds) {
+                handlePatientEvent(patientId, event);
+            }
+        }
+    }
+
+    /**
+     * For the given event, returns the associated patient ids
+     * @param event the event to retrieve the related patient ids
+     * @return the patient ids associated with the given event
+     */
+    public Set<Integer> getPatientIdsForEvent(DbEvent event) {
+        Set<Integer> ret = new HashSet<>();
+        Integer patientId = event.getValues().getInteger("patient_id");
+        if (patientId != null) {
+            ret.add(patientId);
+        }
+        else {
+            for (String key : patientKeys.keySet()) {
+                String keyTable = patientKeys.get(key);
+                Integer value = event.getValues().getInteger(key);
+                if (value != null) {
+                    SqlBuilder sql = new SqlBuilder();
+                    sql.select("p.patient_id").from("patient", "p");
+                    if (keyTable.equals("person")) {
+                        sql.where("p.patient_id = ?");
```

Comment:
Does't this not really do anything?  Or is this just future-proofing for the case when person_id != patient_id?

Reply:
It's needed because there are persons in the DB that are not connected to patients.  So if a change to, say, person_name comes through for the mgoodrich user account, this will not result in an upsert to the dbevent_patient table since there is no associated patient record.

Reply:
Ah right, got it.